pkg/core/bundle_list.go

```diff
@@ -12,52 +13,139 @@ import (
 )
 
 const (
-	maxBundlesToList = 1000000
+	maxMetaFilesToProcess = 1000000
+	typicalBundlesNum     = 1000
 )
 
+func minInt(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// ListBundles returns the list of bundle descriptors from a repo.
+//
+// NOTE: resulting bundles are returned in no particular order.
+//
 // TODO: return a paginated list of id<->bd (separate function in repo_list.go)
-func ListBundles(repo string, store storage.Store) ([]model.BundleDescriptor, error) {
+func ListBundles(repo string, store storage.Store, opts ...BundleListOption) ([]model.BundleDescriptor, error) {
+	settings := CoreSettings{concurrentBundleList: defaultBundleListConcurrency}
+	for _, bApply := range opts {
+		bApply(&settings)
+	}
+
 	// Get a list
 	e := RepoExists(repo, store)
 	if e != nil {
 		return nil, e
 	}
-	ks, _, err := store.KeysPrefix(context.Background(), "", model.GetArchivePathPrefixToBundles(repo), "", maxBundlesToList)
+
+	// this call is not made async yet
+	ks, _, err := store.KeysPrefix(context.Background(), "", model.GetArchivePathPrefixToBundles(repo), "", maxMetaFilesToProcess)
 	if err != nil {
 		return nil, err
 	}
-	bds := make([]model.BundleDescriptor, 0)
-	for _, k := range ks {
+
+	var (
+		workers, wg sync.WaitGroup
+		werr        error
+	)
+
+	bundleChan := make(chan model.BundleDescriptor)
+	keyChan := make(chan string)
+	errorChan := make(chan error)
+
+	for i := 0; i < minInt(settings.concurrentBundleList, len(ks)); i++ {
+		workers.Add(1)
+		go getBundleAsync(repo, store, keyChan, bundleChan, errorChan, &workers)
+	}
+
+	bds := make([]model.BundleDescriptor, 0, typicalBundlesNum) // preallocate some typical number of bundles, e.g. 1000
+
+	wg.Add(1)
+	go func(bundleChan <-chan model.BundleDescriptor, wg *sync.WaitGroup) {
+		defer wg.Done()
+		for bd := range bundleChan { // watch for results and coalesce
+			bds = append(bds, bd)
+		}
+	}(bundleChan, &wg)
+
+	wg.Add(1)
+	go func(errorChan <-chan error, wg *sync.WaitGroup) { // watch for errors
+		defer wg.Done()
+		for err := range errorChan {
+			werr = err
+		}
+	}(errorChan, &wg)
+
+	for _, k := range ks { // distribute work
```

Comment:
was there advantage/disadvantage type reasoning involved in deciding to distribute work synchronously and coalesce results asynchronously?

like, this distribute work loop could probably be included in the goroutine that's waiting for workers to complete?

Reply:
I am avoiding blocking situations in which all workers would be waiting on sending their result and work distributor would be blocked on pushing keys. 

So one of these tasks has to be run in the background. Doing it the other way around is I believe equivalent.

I’ve made some attempts at doing it all in one single loop. This requires the use of select and much extra care with avoiding dropped messages. In the end, this version came up to be much more complex and difficult to read so I decided to drop it and adopt straightforward looping, regardless of possible wasted work whenever a failure occurs.

Reply:
> So one of these tasks has to be run in the background. Doing it the other way around is I believe equivalent.

as the implementation currently exists, i agree: scheduling tasks or coalescing results in the background is equivalent.

> doing it all in one single loop ... this version came up to be much more complex

_totally_ agree here.

----

i kinda suspect that the (for now) equivalent coalescing results synchronously and scheduling tasks in the background will prove to be more durable over time, particularly upon parallelizing the read of keys out of gcs.

yet for the purposes of this patch, it's not worth finagling over in case the review cycle has already become somewhat of a drag.